cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -383,29 +383,65 @@ func generateInitramfsMounts() error {
 	return fmt.Errorf("internal error: mode in generateInitramfsMounts not handled")
 }
 
-func unlockIfEncrypted(name string) (string, error) {
+var (
+	secbootLockAccessToSealedKeys = secboot.LockAccessToSealedKeys
+	osutilFileExists              = osutil.FileExists
+)
+
+// unlockIfEncrypted verifies if an encrypted volume with the specified name exists and unlocks
+// it if this is the case. If this is the last device to be unlocked the access to the sealed keys
+// will be locked. The path to the unencrypted device node is returned.
+func unlockIfEncrypted(name string, last bool) (string, error) {
 	device := filepath.Join("/dev/disk/by-label", name)
 	encdev := device + "-enc"
-	if osutil.FileExists(encdev) {
-		// TODO:UC20: snap-bootstrap should validate that <name>-enc is what
-		//            we expect (and not e.g. an external disk), and also that
-		//            <name> is from <name>-enc and not an unencrypted partition
-		//            with the same name (LP #1863886)
-
-		// TODO:UC20: lock access to keys
-		//            This code only interacts with the TPM if there is an encrypted
-		//            device to unlock, but this misses an important step for when
-		//            there is a TPM but no encrypted device - we must call
-		//            secboot.LockAccessToSealedKeys, and this failing should be
-		//            considered a fatal error. All of the interactions with the TPM
-		//            during boot with the exception of unsealing should happen
-		//            whenever there is a TPM device detected, regardless of whether
-		//            secure boot is enabled or there is an encrypted volume to unlock.
-		sealedKeyPath := filepath.Join(boot.InitramfsEncryptionKeyDir, name+".sealed-key")
-		if err := unlockEncryptedPartition(name, encdev, sealedKeyPath, "", ""); err != nil {
-			return device, fmt.Errorf("cannot unlock %s: %v", name, err)
+
+	// TODO:UC20: use secureConnectToTPM if we decide there's benefit in doing that or we
+	//            have a hard requirement for a valid EK cert chain for every boot (ie, panic
+	//            if there isn't one). But we can't do that as long as we need to download
+	//            intermediate certs from the manufacturer.
+	tpm, tpmErr := insecureConnectToTPM()
+	if tpmErr != nil {
+		logger.Noticef("cannot open TPM connection: %v", tpmErr)
+	} else {
+		defer tpm.Close()
+	}
+
+	var lockErr error
+	err := func() error {
+		defer func() {
+			if last && tpmErr == nil {
+				// Lock access to the sealed keys. This should be called whenever there
+				// is a TPM device detected, regardless of whether secure boot is enabled
+				// or there is an encrypted volume to unlock. Note that snap-bootstrap can
+				// be called several times during initialization, and if there are multiple
+				// volumes to unlock we should lock access to the sealed keys only after
+				// the last encrypted volume is unlocked.
+				lockErr = secbootLockAccessToSealedKeys(tpm)
```

Comment:
@chrisccoulson  is this operation idempotent?  if I read the code correctly if there is an encrypted device we will call this twice, one from Activate* and then explicitly here

Reply:
I verified this with Chris, if an attempt is made to lock the access when it's already locked the second attempt should be ignored (but he can confirm this).